pkg/channel/message_receiver.go

```diff
@@ -149,21 +165,25 @@ func (r *MessageReceiver) ServeHTTP(response nethttp.ResponseWriter, request *ne
 		return
 	}
 
-	// tctx.URI is actually the path...
-	if request.URL.Path != "/" {
-		response.WriteHeader(nethttp.StatusNotFound)
-		return
-	}
-
-	args := ReportArgs{}
-
 	// The response status codes:
 	//   202 - the event was sent to subscribers
 	//   404 - the request was for an unknown channel
 	//   500 - an error occurred processing the request
-	host := request.Host
-	r.logger.Debug("Received request", zap.String("host", host))
-	channel, err := r.hostToChannelFunc(host)
+	args := ReportArgs{}
+	var channel ChannelReference
+	var err error
+
+	// If TLS, use path to determine channel.
+	if eventingtls.IsHttpsSink(request.URL.String()) {
+		channel, err = r.pathToChannelFunc(request.URL.Path)
```

Comment:
If this fails because for example I send a request to `/ns/name/foo`, I'd expect to have a 404 back but we return 404 only when err is UnknownHostError
```
		if _, ok := err.(UnknownHostError); ok {
			response.WriteHeader(nethttp.StatusNotFound)
			r.logger.Info(err.Error())
		}
```

Reply:
I would decouple the receiver itself from checking https/http since this is a library and can be used in different ways and also because a message receiver is basically a server and it either serves http or https requests, never both.

I would do something like this:
- don't configure `pathToChannelFunc` by default
- allow callers to configure `pathToChannelFunc` when creating a message receiver
- if pathToChannelFunc is configured we don't use hostToChannelFunc

in this way, the receiver is either routing requests using host or path but not both based on the scheme and path-based routing can be used for non-https requests too.

Reply:
Okay I like keeping protocol out of the logic. I originally tried to keep just the `hostToChannelFunc` and tried doing both in there if the path was defined in the URL. But it had too many areas for edge cases. I'll update to your suggestion. 

For the `/ns/name/foo`, with the imc usage, the `receiverFunc` would send a `UnknownChannelError` which will then be returned as a 404. The default `hostToChannel/pathToChannel` only sends a generic error. The `UnknownHostError` is there if another implementation overrides the function and returns that error.